[iOS][team-9] 리팩토링 및 뷰 구현

Hansolkkim opened the pull request:
### 🙋🏻 안녕하세요!

안녕하세요! 만사🙋🏻 어느덧 세 번째 PR이네요!
이번주에는 MVVM에 관련된 공부를 진행하고, 저번주에 만사가 남겨주신 피드백 참고해 리팩토링 위주로 진행해보았습니다.
또한 SignInScene의 테스트 코드도 한 번 작성해보았습니다.

부족한 부분이 많지만 리뷰 잘 부탁드리겠습니다!

### **🧑🏼‍💻 작업 목록**

- [x] SignIn Scene의 MVVM구조 리팩토링
- [x] SignIn Scene의 테스트 코드 작성
- [x] BE의 요청으로 인한 로그인 로직 변경
- [x] IssueCollectionView 수정 (HeaderView제거하고 해당 내용을 IssueVC의 NavigationBar로 이동)
- [x] IssueView의 SearchBar가 아랫 방향 스크롤에 의해 나타나도록 구현
- [x] 새로운 이슈를 추가할 때 사용할 EditingIssueView 뷰 구현 및 로직 일부 구현

### **💭 학습 키워드**
- `Navigation Controller`, `Search Contrller`
- `MVVM`
- `XCTest`
- `UISegmentedControl`

### **🤔 고민과 해결**
- SignInViewModel를 여러 추상 타입으로 나누어 구현하려고 했습니다.
    Input, Output, Action의 세 부분으로 나누어 ViewModel을 다시 구현했고,
    DI를 위해 SignInVC를 호출해 사용하는 위치(SceneDelegate)에서 SignInVC 초기화에 필요한 ViewModel, Usecase와 더불어 ViewModel의 Action 객체도 해당 위치에서 생성해 주입해주려고 했습니다.
    여기서 Action 객체는 아래와 같이 정의되어 있었습니다.
    ```swift
    struct SignInViewModelAction {
        let openURL: (URL) -> Void // 로그인 과정에서 외부 Safari 앱을 이용해 URL을 띄워주는 액션
        let presentTabBarController: () -> Void // 로그인이 완료되면 다음 화면인 TabBarController로 넘어가는 액션
    }
    ```
    하지만 제가 사용하려고 한 액션들은 SceneDelegate에서 정의해주기가 어려운 액션이었습니다.(TabBarController의 경우 SceneDelegate에서 생성해주는 것이 어색했고, 더군다나 SceneDelegate에는 `present(_:animated:)` 메소드가 없어 다음 VC를 띄워주는 액션을 정의하기가 불가능했습니다.)
    
    그래서 먼저, TabBarController를 띄워주는 액션은 SignInViewModel의 액션에서 제거했습니다. 로그인이 완료된 후 앱으로 redirect되면 SceneDelegate가 가장 먼저 호출되기 때문에, 로그인이 됐을 경우 rootViewController를 TabBarController로 설정하여 화면이 넘어가도록 했습니다.
    
    그리고 `openURL` 액션의 경우에는 SignInVC에서 ViewModel과 View를 binding을 해주는 메소드 내에서 설정해주도록 했습니다. 굳이 SceneDelegate에서 해당 액션을 지정해서 ViewModel의 의존성을 주입해줄 필요성이 없다고 판단했기 때문입니다.

- 새로운 Issue를 작성할 때 내용(Contents)에 해당하는 부분을 `UITextView`를 이용해서 구현하려 했습니다.
    여러 줄을 보여주고, 편집할 수 있는 UI로는 TextField보다 TextView가 더 적합하다고 판단했기 때문입니다.
    하지만 TextField와 달리, TextView에는 `placeholder`를 설정할 수 있는 속성이 없어 어떻게 할지 고민했습니다.
    
    이를 위해, UITextViewDelegate를 이용해, `해당 TextView에 편집이 되기 전 상태 + 단순 빈칸으로만 Text가 작성된 상태` 에서는 placeholder에 해당하는 문구를 회색 폰트로 보여주도록 했고,
    
    편집이 시작되고 글자가 Text에 입력된 경우에는 placeholder에 해당하는 문구를 지우고 검은색 폰트로 보여주도록 하여, placeholder를 UITextView에서도 나타내주도록 했습니다.
    구현한 Delegate 코드는 아래와 같습니다!
    ```swift
    extension EditingIssueViewController: UITextViewDelegate {
        private var placeholder: String {
            return "코멘트는 여기에 작성하세요."
        }

        func textViewDidBeginEditing(_ textView: UITextView) {
            if textView.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                textView.textColor = .gray
                textView.text = placeholder
            } else if textView.text == placeholder {
                textView.textColor = .black
                textView.text = nil
            }
        }

        func textViewDidEndEditing(_ textView: UITextView) {
            if textView.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                textView.textColor = .gray
                textView.text = placeholder
            }
        }
    }
    ```

- EditingIssueView에서 상단의 SegmentedControl이 선택된 index에 따라 마크다운, 미리보기 뷰가 전환되도록 하는 기능을 구현해야 했습니다.
    이를 구현하기 위해 해당하는 위치를 View로 만들어놓고 각 경우마다 해당 위치에 들어갈 View를 바꿔주려고 하다가,
    이렇게 하면 각각에 들어갈 Text를 계속해서 서로에게 넘겨줘야하는 문제가 있어, 원본 Text를 따로 변수로 저장해놓아야하는 불필요한 과정이 생겼습니다.
    그래서 두 View를 같은 위치에 위치시키고, View의 alpha를 이용해 각 경우에 해당하는 View만 `alpha = 1` 로 설정해 보여지도록 구현했습니다. 해당 로직은 아래와 같이 구현했습니다.
    ```swift
    // EditingIssueView
    func changeContentView(to selectedIndex: Int) {
        let unselectedIndex = (selectedIndex == 0) ? 1 : 0
        let contentViews = [contentTextView, previewTextView]

        if contentTextView.text != "코멘트는 여기에 작성하세요." {
            previewTextView.attributedText = SwiftyMarkdown(string: contentTextView.text).attributedString()
        }

        UIView.animate(withDuration: 0.5) {
            contentViews[selectedIndex].alpha = 1
            contentViews[unselectedIndex].alpha = 0
        }
    }
    ```



### **💬  실행 화면**

|IssueView의 SearchBar가 스크롤 됐을 때 나타나는 화면|EditingIssueView 화면|
|---|---|
|![SS 2022-06-22 PM 05 35 44](https://user-images.githubusercontent.com/92504186/174983698-05370ba3-a5cb-4b9c-9ab3-c9795527b87d.gif)|![SS 2022-06-22 PM 05 40 30](https://user-images.githubusercontent.com/92504186/174984741-a3e118db-42b6-45bd-b5c7-82fa7082a0d9.gif)|

### **🙋🏻 질문 거리**
- 처음에 SignInViewModel의 추상 타입인 SignInViewModelProtocol은 SignInViewModelInput, SignInViewModelOutput 두 프로토콜을 동시에 채택하는 Protocol 이었습니다. 이를 위해 해당 SignInViewModelProtocol은 안에 아무런 내용 없이 아래와 같이 정의했습니다.
    ```swift
    protocol SignInViewModelInput {
        func didSelect(_:)
    }
    protocol SignInViewModelOutput {
        var something: Observable<T>
    }
    protocol SignInViewModelProtocol: SignInViewModelInput, SignInViewModelOutput { }
    ```
    하지만 이렇게 아무런 내용없이 Protocol을 하나 만드느니, typealias를 이용해 그저 이름만 가지도록 하는 방법을 사용하는게 어떨까 싶어 아래와 같이 SignInViewModelProtocol을 정의해봤고, 두 방법에 의한 동작의 차이는 없었습니다.
    ```swift
    typealias SignInViewModelProtocol = SignInViewModelInput & SignInViewModelOutput
    ```
    
    제가 생각한 아래 방법은 Protocol 추상 타입을 정의하는데 있어서 잘 사용되지 않는 방법일까요?
    typealias를 사용해서 만든 것이 Protocol이라는 네이밍을 갖는 것이 어색해서 궁금합니다..ㅎㅎ

Hansolkkim committed:
refactor: StubURLSession.swift에 함께 정의되어 있던 Mock 관련 객체들 모두 분리 (bugpigg/issue-tracker#16)

JasonLee0223 committed:
refactor: IssueCollectionDataSource 파일의 data[indexPath] 부분 수정

JasonLee0223 committed:
feat: IssueCollectionDataSource파일의 header if-else구문 Switch구문으로 변경

Hansolkkim committed:
feat: Observable<T> 타입 선언 (bugpigg/issue-tracker#16)

action property들을 private한 프로퍼티로 설정하고 setter를 위한 메소드를 제공할 수 있는 타입으로 사용할 수 있도록 Observable 타입 선언

Hansolkkim committed:
refactor: SignInViewModel의 Observer 속성들의 타입을 Observable 타입으로 변경 (bugpigg/issue-tracker#16)

Hansolkkim committed:
rename: NetworkTargetable -> NetworkTargetProtocol/ NetworkManager -> NetworkService 이름 변경 (bugpigg/issue-tracker#16)

NetworkTargetProtocol의 경우, 어색한 네이밍으로 인해 이름 변경
NetworkManager의 경우, Manager라는 이름을 usecase에서도, service에서도 사용하여 발생할 혼란을 방지하기 위해 이름 변경

Hansolkkim committed:
refactor: UIColor의 너무 긴 이니셜라이저의 길이를 줄이기 위한 static 메소드들 선언 (bugpigg/issue-tracker#16)

JasonLee0223 committed:
refactor: a5110ac 커밋 내용 코멘트 다시 적용

Hansolkkim committed:
feat: URLSessionProtocol에 dataTask 메소드 추가 (bugpigg/issue-tracker#16)

NetworkService의 fetchData 메소드에서 사용하는 URLRequest를 이용해 dataTask를 만드는 메소드를 URLSessionProtocol에 추가

Hansolkkim committed:
feat: fetchData 메소드에서 사용할 URLSession을 매개변수로 받아올 수 있도록 매개변수 수정 (bugpigg/issue-tracker#16)

Hansolkkim committed:
refactor: NetworkService 객체가 URLSessionProtocol 타입 프로퍼티를 가지고 있도록 수정 (bugpigg/issue-tracker#16)

NetworkService 테스트시 StubURLSession을 사용할 수 있도록, 프로퍼티로 URLSessionProtocol을 가지고 있도록 수정

Hansolkkim committed:
feat: BE 요청으로 인한 JWT Token 요청 로직 변경 (bugpigg/issue-tracker#16)

Hansolkkim committed:
chore: JWT 토큰을 요청하는 URL 수정 (bugpigg/issue-tracker#16)

JasonLee0223 committed:
feat: headerView에 searchBar 추가

Hansolkkim committed:
test: SignInManager 테스트 코드 작성 완료 (bugpigg/issue-tracker#20)

Hansolkkim committed:
test: NetworkService 테스트 코드 작성 완료 (bugpigg/issue-tracker#20)

Hansolkkim committed:
fix: SignInVC bind(to:) 메소드 호출 위치로 인한 이슈 해결

- bind 메소드가 SignInView가 생기기 전에 호출되어 button과 viewModel 간 bind가 되지 않아서 버튼 터치 이벤트가 발생하지 않는 이슈 해결

Hansolkkim committed:
refactor: SingInViewModel Test를 위해 SignInManagable 프로토콜 구현 (bugpigg/issue-tracker#20)

- SignInViewModel Test시 MockSignInManager를 사용하기 위해 SignInManager가 채택할 SignInManagable 프로토콜 구현

Hansolkkim committed:
test: SignInViewModel 테스트 코드 작성 완료 (bugpigg/issue-tracker#20)

Hansolkkim committed:
Merge remote-tracking branch 'origin/iOS-test/SignInScene' into iOS

JasonLee0223 committed:
feat: searchBar 생성

Hansolkkim committed:
chore: UIColor private 메소드, 열거형 이름 변경

- colur(from:colur:) -> getColor(from:colorType:)
- Colour -> ColorType

Hansolkkim committed:
refactor: SignInViewModel의 구조 변경

- SignInViewModel의 Input, Output에 해당하는 속성, 메소드를 프로토콜로 나누어서 정의
- 의존성 역전을 위해 SceneDelegate에서 SignInVC의 속성들을 정해주는 메소드 추가
- SignInViewModel의 애매했던 프로퍼티 이름 정리 및 Observing 하지 않는 프로퍼티, action과 중복되는 프로퍼티 제거 (errorAction -> error/ presentAction, buttonAction 제거)
- SignInViewModelActions의 presentTabBarController 속성의 값을 SceneDelegate에서 설정해주려고 하니 SceneDelegate에는 present() 메소드가 존재하지 않아, 어쩔 수 없이 SignInViewModel에서 해당 속성을 set해주는 메소드를 제공 (setPresentAction(_:))

Hansolkkim committed:
test: SignInViewModel 구조 변경으로 인한 테스트 수정 (bugpigg/issue-tracker#20)

Hansolkkim committed:
refactor: SceneDelegate에서 TabBarController를 띄워주도록 로직 변경 (bugpigg/issue-tracker#21)

- 해당 로직이 변경됨에 따라 SignInViewModelActions에서 presentTabBarController 프로퍼티 제거
- 위의 이유로 SceneDelegate에서 SignInViewModel로 로그인 완료에 대한 노티를 날리는 부분 삭제
- 기존의 로직으로는 SignInVC가 화면에서 사라져도 살아있는 문제가 있어, SceneDelegate에서 UINavigationController를 이용해 상황에 맞는 VC를 띄워주도록 로직 변경
- Actions의 의존성을 주입해주는 부분을 SignInVC로 옮김에 따라 SceneDelegate에서 SignInVC를 생성할 때 ViewModelAction의 의존성 주입해주는 부분 제거
- 위의 이유로 SignInViewModel에서 Actions의 openURL 프로퍼티의 값을 설정해주는 메소드 제공하도록 구현

Hansolkkim committed:
refactor: SignInManager의 requestJWTToken() 메소드에서 code를 다루는 로직 수정 (bugpigg/issue-tracker#21)

- code를 String이 아닌 URLComponents로 다루도록 로직 수정

JasonLee0223 committed:
refactor: searchBar collectionViewHeader에서 navigationItem으로 옮기는 수정

JasonLee0223 committed:
Merge branch 'iOS-#Refactor' into iOS
iOS-#Refactor 브랜치 머지 (Jason)

Hansolkkim committed:
fix: UINavigationViewController가 두 개 생성되는 이슈 해결 (bugpigg/issue-tracker#21)

- SceneDelegate에서 SignInVC의 메모리 해제를 위해 NavigationController에 담아줬던 이유로 발생헀던 이슈 해결

Hansolkkim committed:
feat: 새로운 이슈 추가하는 버튼 구현 (bugpigg/issue-tracker#22)

- 새로운 이슈 추가 버튼 뷰 구현
- 해당 버튼에 Action을 추가하는 메소드 구현
- 해당 버튼의 backgroundColor로 이용할 색상을 UIColor Extension을 이용해 추가

JasonLee0223 committed:
feat: SearchBar 설정 및 스크롤 확인을 위한 dummydata 추가

JasonLee0223 committed:
feat: 스크롤이 끝나면 서치바가 나오게끔 변경

Hansolkkim committed:
refactor: CustomBarButton을 재사용 가능하도록 수정 (bugpigg/issue-tracker#22)

Hansolkkim committed:
feat: EditingIssueView 구현 (bugpigg/issue-tracker#22)

- 새로운 이슈를 생성하거나, 이슈를 수정할 때 사용할 EditingIssueView 구현
- ContentTextView의 placeholder 설정을 위해 EditingIssueVC에서 UITextViewDelegate 채택
- TitleTextField, ContentTextView의 text 유무에 따라 SaveButton 활성화 유무 로직 구현 (UITextFieldDelegate, UITextViewDelegate 이용)
- 임시적으로 취소 버튼의 액션 구현 (추후 ViewModel 구현해서 수정할 예정)
- IssueVC에서 plusButton을 터치했을 때 EditingIssueVC가 화면에 나타나는 로직 구현

Hansolkkim committed:
fix: SignInViewModel의 actions을 초기화시켜주지 않아서 발생한 이슈 해결

Hansolkkim committed:
chore: IssueVC searchBar 관련 로직 일부 수정

Hansolkkim committed:
feat: SwiftyMarkdown pod 이용해 마크다운 미리보기 뷰 구현 (bugpigg/issue-tracker#22)

- 'SwiftyMarkdown' pod 추가
- EditingIssueView에 미리보기 뷰 추가
- 해당 미리보기 뷰와 contentTextView의 alpha를 이용해 보여주려는 뷰 변경해주는 로직 구현
- EditingIssueVC의 selectContentViewSegment에 Action 추가
- EditingIssueView에서 selectContentViewSegment가 바뀜에 따라 미리보기 뷰, 마크다운 뷰를 보여주는 메소드 제공
- titleTextField, contentTextView의 속성 변경 (자동 수정, 자동 첫글자 대문자 속성 변경)

Hansolkkim committed:
Merge remote-tracking branch 'origin/iOS-feat/IssueSceneView' into iOS

# Conflicts:
#	iOS/IssueTracker/IssueTracker/Issue/View/IssueCollectionView.swift
#	iOS/IssueTracker/IssueTracker/Issue/View/IssueViewController.swift

Hansolkkim committed:
chore: merge로 인한 코드 문제 해결

Hansolkkim committed:
test: SignInViewModel 변화로 인한 테스트 코드 변경

ITzombietux reviewed (approved):
안녕하세요!! 오늘 하루도 대단히 수고많으셨습니다!!
코드를 보면서 전반적으로 가독성이 꽤 높아서 읽기가 좋았네요!
메서드 네이밍, 길이, 뎁스 같은 부분이 좋았던거 같아요!! 
UI 구현하신 부분은 워낙 잘해주셔서 제가 크게 리뷰할게 없어 보였어요!
이후엔는 사소한 segment라던지 하는 부분도 하위 객체로 나누어 보면 좋을거 같네요!

테스트 코드도 잘 적어주신거 같고, 이후에 커버리지를 점점 높여가는 것도 즐거운 재미가 있을거 같네요!

- 질문에 대한 저의 답변은, 아무런 내용이 없으면 사실 굳이 만들 필요는 없다고 생각이 됩니다.! 이것도 두 분이 어떻게 정하느냐에 따라 다를거 같은데요! 통일서 있게 하자! 라고 한다면 역할이 없더라도 만들 수 있을테고, 유연하게 가자고 한다면 없애겠죠!! ㅎㅎㅎ typealias를 사용한건 좋은 생각이라고 생각합니다만! Protocol이라는 단어가 붙는건 어색하긴 하네요!! ㅎㅎㅎ -Type 이라던지 뭐 이런 식으로 해도 될거 같습니다 ㅎㅎ

Hansolkkim committed:
rename: cellForItemAt에서 사용하는 변수명 수정 (bugpigg/issue-tracker#26)

- items -> issueEntity로 수정
- cell의 configuration을 결정해주는 데이터는 Entity라고 생각하여, issueEntity로 수정

Hansolkkim committed:
chore: [String: String]을 typealias 이용, "StringDictionary" 로 정의 (bugpigg/issue-tracker#26)

Hansolkkim committed:
chore: 코드 일부 수정 (bugpigg/issue-tracker#26)

- SignInManager의 urlSession 속성 접근 제어자 지정 (private let)
- NetworkServiceTest의 print문 모두 제거

Hansolkkim committed:
refactor: EditingIssueVC와 IssueVC에 있는 NavigationItem 뷰들의 정의를 VC로부터 분리 (bugpigg/issue-tracker#26)

Hansolkkim commented on iOS/IssueTracker/IssueTracker/Issue/View/EditingIssueViewController.swift:
```diff
@@ -0,0 +1,114 @@
+}
+
+// MARK: - UITextFieldDelegate for enabling Save Button
+extension EditingIssueViewController: UITextFieldDelegate {
```
해당 부분 수정 진행하다가, 궁금한 점이 생겼습니다!
TextFieldDelegate의 메소드 중 `textFieldDidBeginEditing(_:)` 메소드에서는 EditingIssueView의 titleTextField.text 뿐아니라, ContentTextView.text 또한 확인합니다.
TextFieldDelegate를 VC로 부터 분리하다보니, View로부터 해당 정보를 받아오는게 복잡해졌는데, 이런 경우에는 어떤 방식으로 VC 및 View <-> Delegate 객체 간 통신을 하는게 적합할까요?

Hansolkkim replied:
MVVM 구조로 개선하다가 깨달았는데, ViewModel에서 두 text를 모두 observing하고 있으면 어느 정도 통신의 문제는 해결할 수 있을 것 같습니다..ㅎㅎ

Hansolkkim replied:
MVVM 구조로 개선 후 궁금증이 생겼습니다!
Delegate 객체를 VC로부터 분리를 하다보니, VC가 Delegate의 메소드에 대한 응답을 받고, 이를 ViewModel로 전달해주는 흐름이 생겼습니다. (예를 들어, Delegate 객체에서 textViewDidChanged() 메소드가 호출되면 Delegate 객체에서는 이를 VC에게 알려주고, VC는 이 알림을 받으면 ViewModel의 Observable 프로퍼티의 value를 바꿔줍니다.)

이렇게 복잡한 흐름을 구현하는 것보다, Delegate를 객체로 분리하지 않고 VC가 Delegate 역할을 맡고, VC가 delegate 메소드가 호출 될 때 ViewModel에게 알려주는게 흐름상 보기 좋지 않을까 생각이 들었습니다.

혹시 만사의 생각은 어떠신지 궁금합니다!

Hansolkkim commented on iOS/IssueTracker/IssueTracker/SignIn/Model/SignInManager.swift:
```diff
@@ -7,10 +7,23 @@
 struct SignInManager {
     private let clientID = Bundle.main.clientID
     private let clientSecret = Bundle.main.clientSecret
+    var urlSession: URLSessionProtocol
```
접근 제어자를 왜 안붙였는지 저도 모르겠네요..ㅋㅋㅋㅋ 
반영하고 커밋하겠습니다!

Hansolkkim commented on iOS/IssueTracker/IssueTracker/Issue/View/EditingIssueViewController.swift:
```diff
@@ -0,0 +1,114 @@
+import SnapKit
+
+final class EditingIssueViewController: UIViewController {
+    private let selectContentViewSegment: UISegmentedControl = {
+        let segment = UISegmentedControl(items: ["마크다운", "미리보기"])
+        segment.selectedSegmentTintColor = .white
+        segment.selectedSegmentIndex = 0
+        return segment
+    }()
+
+    private let saveButton: CustomBarButton = {
+        let customButton = CustomBarButton()
+        customButton.setConfiguration(title: "저장", imageSystemName: "plus", imagePlacement: .trailing)
+        return customButton
+    }()
+
+    private let cancelButton: CustomBarButton = {
+        let customButton = CustomBarButton()
+        customButton.setConfiguration(title: "취소", imageSystemName: "chevron.backward", imagePlacement: .leading)
+        return customButton
+    }()
```
말씀 주신대로, NavigationItem으로 사용될 View들의 정의를 다른 구조체에서 하도록 해서 사용했습니다.
그런데 NavigationItem이기 때문에, 따로 Constraint를 설정해줄 필요가 없어, 그저 구조체로 뷰들의 정의만 옮긴 정도인데,
이렇게 분리를 하는 것도 괜찮은 방법일까요?
아래와 같이 분리했습니다.
```swift
struct EditingIssueViewNavigationItems {
    let selectContentViewSegment: UISegmentedControl = { ... }()

    let saveButton: CustomBarButton = { ... }()

    let cancelButton: CustomBarButton = { ... }()
}

// EditingIssueViewController
let navigationItems = EdiginIssueViewNavigationItems()

self.navigationItem.leftButton = UIBarButtonItem(customView: navigationItems.cancelButton)
```

ITzombietux replied:
너무 좋습니다!! ㅎㅎㅎㅎ

Hansolkkim committed:
refactor: EditingIssueView의 컴포넌트들을 분리 (bugpigg/issue-tracker#26)

- title, content를 기준으로 EditingIssueView의 컴포넌트를 분리
- 컴포넌트로 분리하여 View를 위한 코드가 줄어들어, EditingIssueView를 제거하고 뷰 요소들을 EditingIssueVC가 가지도록 수정
- UIStackView를 extension하여, 여러 뷰를 한 번에 addArrangeSubview 하도록 메소드 추가

Hansolkkim commented on iOS/IssueTracker/IssueTracker/Issue/View/EditingIssueView.swift:
```diff
@@ -0,0 +1,155 @@
+import SwiftyMarkdown
+
+final class EditingIssueView: UIView {
+    private let titleStackView: UIStackView = {
+        let stackView = UIStackView()
+        stackView.translatesAutoresizingMaskIntoConstraints = false
+        stackView.axis = .horizontal
+        stackView.distribution = .fillProportionally
+        return stackView
+    }()
+
+    private let titleLabel: UILabel = {
+        let label = UILabel()
+        label.translatesAutoresizingMaskIntoConstraints = false
+        label.text = "제목"
+        label.font = UIFont(name: "SFProDisplay-Regular", size: 17)
+        return label
+    }()
+
+    private let titleTextField: UITextField = {
+        let textField = UITextField()
+        let screenSize = UIScreen.main.bounds.size
+        textField.translatesAutoresizingMaskIntoConstraints = false
+        textField.font = UIFont(name: "SFProDisplay-Regular", size: 17.0/812.0 * screenSize.height)
+        textField.alpha = 1
+        textField.autocorrectionType = .no
+        textField.autocapitalizationType = .none
+        return textField
+    }()
```
말씀하신 방법 사용해서, title, content를 기준으로 컴포넌트를 나누어 적용해보았습니다. 
View Components를 나누다보니 EiditingIssueView에서의 뷰에 대한 코드 길이가 많이 줄어, 굳이 View를 ViewController로부터 나누고, View가 ViewComponents를 가지고 있을 필요가 없다고 생각하여, 뷰 관련 코드도 모두 EditingIssueVC가 가지도록 수정했습니다!

Hansolkkim committed:
refactor: EditingIssueScene이 MVVM 구조를 가지도록 변경 (bugpigg/issue-tracker#26)

- EditingIssueViewModel 구현
- EditingIssueVC와 EditingIssueViewModel 바인딩
- EditingIssueVC로 부터 TextFieldDelegate, TextViewDelegate 분리
- Notification을 이용해 Delegate와 VC간 통신 구현

Hansolkkim committed:
feat: EditingIssueManager 구현 (bugpigg/issue-tracker#26)

- EditingIssueScene의 usecase로 사용할 EditingIssueManager 구현
- EditingIssueManager에서 네트워크 통신시 사용할 EditingIssueNetworkTarget 구현

Hansolkkim committed:
feat: save버튼 및 에러에 대한 Observable 프로퍼티 구현 및 usecase 프로퍼티 추가 (bugpigg/issue-tracker#26)

Hansolkkim committed:
feat: post에 대응해 httpBody를 추가하는 코드 추가 (bugpigg/issue-tracker#26)

Hansolkkim committed:
test: EditingIssueManager 객체에 대한 테스트 코드 작성 (bugpigg/issue-tracker#26)

- 테스트를 위해 해당 객체의 urlSession 속성을 설정해줄 필요가 있어, 해당 속성을 변수로 바꾸고, 해당 속성을 설정해주는 메소드 추가 구현

Hansolkkim committed:
Revert "test: EditingIssueManager 객체에 대한 테스트 코드 작성 (bugpigg/issue-tracker#26)"

This reverts commit f895c0de3838cfae526af606e1e65287b014db30.

Hansolkkim committed:
Revert "feat: post에 대응해 httpBody를 추가하는 코드 추가 (bugpigg/issue-tracker#26)"

This reverts commit c1ae9fa89816874c6687e840bd9c896ccfd15707.

Hansolkkim committed:
Revert "feat: save버튼 및 에러에 대한 Observable 프로퍼티 구현 및 usecase 프로퍼티 추가 (bugpigg/issue-tracker#26)"

This reverts commit 36bbcbccc83514c56a4aafd5395569fd19d9b6f2.

Hansolkkim committed:
Revert "feat: EditingIssueManager 구현 (bugpigg/issue-tracker#26)"

This reverts commit 1fa2f257750653d9748424e33d3be50b22c7890d.
